### PR TITLE
Security verification

### PR DESCRIPTION
**Pull request by younggentech**
Please review the security code for the application.

### CODE REVIEW

File: src/homework/api/security.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        secret,
+        algorithm="HS256",
+    )
+
+
+def verify_token(token: str, supposed_secret: str) -> typing.Dict[str, str]:
+    return jwt.decode(token, supposed_secret, algorithms="HS256")
+
+
+def get_app_by_id(
```
Comment (ilystsov) [inline src/homework/api/security.py]: What do you thing about moving all database-related functions into a separate file?

File: src/homework/api/security.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    """
+    engine = create_engine(get_url_func())
+    with Session(engine) as session:
+        app = session.scalar(
+            select(Application).where(Application.app_id == app_id)
+        )
+        return app
+
+
+def decode_segment(segment: str) -> None | typing.Dict[str, str]:
```
Comment (ilystsov) [inline src/homework/api/security.py]:
```suggestion
def decode_segment(segment: str) -> None | dict[str, str]:
```
This suggestion may enhance readability, although it is very minor

File: src/homework/api/security.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-in-path-operation-decorators/
+    Specify security level for a router:
+    >>> from fastapi import APIRouter, Depends
+    >>> read_only_access_level = AccessLevel("Can_Read")
+    >>> router = APIRouter(
+    >>>     prefix="/items",
+    >>>     tags=["items"],
+    >>>     dependencies=[Depends(ValidateHeader(read_only_access_level))],
+    >>>     responses={404: {"description": "Not found"}},
+    >>>     )
```
Comment (ilystsov) [inline src/homework/api/security.py:94]: Great job on the docstrings! Including both use case and a reference to FastApi docs is an excellent idea.

File: src/homework/api/security.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        it returns {"detail": ...} json instead of {"success": False}
+        (priority - low)
+    """
+
+    def __init__(self, required_access_level: BroadenAccessLevel):
+        self.required_access_level = required_access_level
+
+    async def __call__(self, token: typing.Annotated[str, Header()]):
+        if not token_has_access(token, self.required_access_level):
+            raise HTTPException(status_code=403)
```
Comment (ilystsov) [inline src/homework/api/security.py:106]: Using `dependencies` this way is another good idea!

**Review by ilystsov — COMMENTED** [review]
Overall great job on tests and code.

File: src/homework/api/security.py (hunk 1 above)
Comment (younggentech) [reply src/homework/api/security.py]: Ok, this is a good point

File: src/homework/api/security.py (hunk 2 above)
Comment (younggentech) [reply src/homework/api/security.py]: Great idea, thanks! 
P. S. the idea works for Python3.9 and upper, which is suitable for us :)
https://stackoverflow.com/a/68908479

**Pull request merged**